src/backend/gl/src/device.rs

```diff
@@ -1096,51 +1127,91 @@ impl d::Device<B> for Device {
 
     unsafe fn unmap_memory(&self, memory: &n::Memory) {
         let gl = &self.share.context;
-        let buffer = match memory.first_bound_buffer.get() {
-            None => panic!("No buffer has been bound yet, can't map memory!"),
-            Some(other) => other,
-        };
-        let target = glow::PIXEL_PACK_BUFFER;
+        let buffer = memory.buffer.expect("cannot unmap DEVICE_LOCAL memory");
 
-        gl.bind_buffer(target, Some(buffer));
+        gl.bind_buffer(memory.target, Some(buffer));
 
         if self.share.private_caps.emulate_map {
-            let raw = memory.emulate_map_allocation.replace(std::ptr::null_mut());
-            let mapped = slice::from_raw_parts_mut(raw, memory.size as usize);
+            let ptr = memory.emulate_map_allocation.replace(None).unwrap();
+            let allocation = slice::from_raw_parts_mut(ptr, memory.size as usize);
             // TODO: Access
-            gl.buffer_data_u8_slice(target, mapped, glow::DYNAMIC_DRAW);
-            let _ = *Box::from_raw(raw);
+            gl.buffer_sub_data_u8_slice(memory.target, 0, &allocation);
```

Comment:
we shouldn't be doing this on `unmap`. In Vulkan and gfx-hal, mapping as an operation doesn't actually involve any synchronization. I.e. the only thing that synchronizes is:
  - submit() for coherent memory (implicitly synchronizes all active coherent mappings)
  - explicit flush() and invalidate() for non-coherent memory

Reply:
That makes sense, thanks for being patient with me and helping me learn all of the different memory semantics for modern gfx apis!